org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java

```diff
@@ -0,0 +1,239 @@
+      result = 31 * result + (line != null ? line.hashCode() : 0);
+      return result;
+    }
+  }
+
+  private enum LineAndLineHashKeyFactory implements SearchKeyFactory {
+    INSTANCE;
+    @Override
+    public SearchKey create(Trackable t) {
+      return new LineAndLineHashKey(t);
+    }
+  }
+
+  private static class LineHashAndMessageKey implements SearchKey {
+    private final String ruleKey;
+    private final String message;
+    private final Integer lineHash;
+
+    LineHashAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.lineHash = trackable.getLineHash();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      LineHashAndMessageKey that = (LineHashAndMessageKey) o;
+      // start with most discriminant field
+      return Objects.equals(lineHash, that.lineHash)
+        && message.equals(that.message)
+        && ruleKey.equals(that.ruleKey);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = ruleKey.hashCode();
+      result = 31 * result + message.hashCode();
+      result = 31 * result + (lineHash != null ? lineHash.hashCode() : 0);
+      return result;
+    }
+  }
+
+  private enum LineHashAndMessageKeyFactory implements SearchKeyFactory {
+    INSTANCE;
+    @Override
+    public SearchKey create(Trackable t) {
+      return new LineHashAndMessageKey(t);
+    }
+  }
+
+  private static class LineAndMessageKey implements SearchKey {
+    private final String ruleKey;
+    private final String message;
+    private final Integer line;
+
+    LineAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.line = trackable.getLine();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      LineAndMessageKey that = (LineAndMessageKey) o;
+      // start with most discriminant field
+      return Objects.equals(line, that.line)
+        && message.equals(that.message)
+        && ruleKey.equals(that.ruleKey);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = ruleKey.hashCode();
+      result = 31 * result + message.hashCode();
+      result = 31 * result + (line != null ? line.hashCode() : 0);
+      return result;
+    }
+  }
+
+  private enum LineAndMessageKeyFactory implements SearchKeyFactory {
+    INSTANCE;
+    @Override
+    public SearchKey create(Trackable t) {
+      return new LineAndMessageKey(t);
+    }
+  }
+
+  private static class LineHashKey implements SearchKey {
+    private final String ruleKey;
+    private final Integer lineHash;
+
+    LineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.lineHash = trackable.getLineHash();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Add a type test to this method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)